### PR TITLE
[c4core/ryml] Fix gcc4.8 compilation errors and update misc

### LINKED ISSUE #29876: [c4core] Need to include compat headers
****

**Is your feature request related to a problem? Please describe.**
This library is used by the ryml library. c4core offers backwards compat to gcc 4.8. However, the necessary files are not being picked up the port.  The result is that compilation with the older compiler is not possible as the necessary headers cannot be found.

**Proposed solution**
c4core port needs to pick the files located in the cmake submodule used by the project. 

**Describe alternatives you've considered**
None.

**Additional context**
Found this problem trying to use vcpkg with makefiles (not cmake) on an existing project that compiles against gcc 4.8. We added ryml and found that compilation could not complete.

### PR DESCRIPTION
**Pull request by jimwang118**
Fixes https://github.com/microsoft/vcpkg/issues/29876
```
/home/vsts/work/1/vcpkg/installed/x64-linux/include/c4/compiler.hpp:105:41: fatal error: c4/gcc-4.8.hpp: No such file or directory
 #               include "c4/gcc-4.8.hpp"
```
Update the content in the `cmake` folder and add the include path of the `gcc-4.8.hpp` header file.
- [X] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [X] SHA512s are updated for each updated download
- [ ] ~~The "supports" clause reflects platforms that may be fixed by this new version~~
- [ ] ~~Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.~~
- [ ] ~~Any patches that are no longer applied are deleted from the port's directory.~~
- [X] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [X] Only one version is added to each modified port's versions file.

Usage test pass with following triplets:

```
x64-linux
```

### CODE REVIEW

**Comment by dg0yt** [comment]
> I'm curious why the style of individually/manually downloading submodules is used?

It makes it possible to use asset caching, or even completely disconnected builds. And most of the time, de-vendoring is the preferred approach.
FWIW, if the refs can be read from a file, it should be possible to add validation/warnings/errors to the portfile. (Warnings are used in the skia port.)

**Comment by jimwang118** [comment]
> I'm curious why the style of individually/manually downloading submodules is used? Wouldn't this be easier if `vcpkg_execute_required_process` was used instead? Then all you have to manage is 1 commit moving forward and all submodules are picked up correctly (ref [#6886 (comment)](https://github.com/microsoft/vcpkg/issues/6886#issuecomment-502617274)).
> 
> Allowing git to pull in the submodules is a better approach for c4core and especially ryml, which relies on c4core. As it is currently provisioned ryml is still incorrect.If you download the 3rd tier submodule and bypassing the 2nd, then don't you risk missing changes made in the 2nd tier (or other 3rd tier subs)?
> 
> Desired: ryml -> c4core -> cmake, debugbreak,fastfloat Actual: ryml -> (skip c4core) -> cmake (skip debugbreak, fastfloat)
> 
> The ryml library is incomplete and if there are any changes in c4core, debugbreak or fastfloat, those changes would never make it into the stated version.

We wrote c4core in the dependencies of ryml, and then used find_package (c4core CONFIG REQUIRED) to call the installed c4core library, so as long as c4core updates cmake, debugbreak, and fastfloat, they will be applied in the ryml library.

**Comment by mprather** [comment]
@jimwang118 : Thanks for the info on how the ryml ties into c4core with respect to vcpkg.  Isn't there a problem if c4core has a higher version # than the ryml version if vcpkg.json doesn't specify a version? The two project can move independently, right? Shouldn't ryml's vcpkg designate a version for c4core?

@dg0yt : It's clear I have more to learn about vcpkg port authoring as I didn't even know there was a concept of disconnected builds. If I look at the documentation on `vcpkg_from_github` and `vcpkg_download_distfile` I don't see how these functions would be "more disconnectable" vs `vcpkg_execute_required_process`. The docs don't offer any indication why I would use one over the other. Is there a blog post or other reference you can recommend on authoring ports?

RE: putting refs in a file --- isn't that redundant and more error prone than actually using the info already baked into submodules? Or am I misunderstanding the suggestion? I took a look at the skia project and doesn't use submodules. I'm not connecting the dots with the suggestion.

**Comment by jimwang118** [comment]
> @jimwang118 : Thanks for the info on how the ryml ties into c4core with respect to vcpkg. Isn't there a problem if c4core has a higher version # than the ryml version if vcpkg.json doesn't specify a version? The two project can move independently, right? Shouldn't ryml's vcpkg designate a version for c4core?
> 
You are right, the c4core version that ryml needs to depend on is fixed. Now the updated result is that the version of c4core is the version that ryml needs. Now the version dependency is correct. If updated later, the two need to be updated together to ensure that the c4core version required by ryml is correct.

**Review by MonicaLiu0311 — DISMISSED** [review]

File: ports/c4core/portfile.cmake [hunk 1]
```diff
@@ -7,16 +7,18 @@ vcpkg_from_github(
...
+        fix_gcc_48.patch
 )
 
-set(CM_COMMIT_HASH fe41e86552046c3df9ba73a40bf3d755df028c1e)
+set(CM_COMMIT_HASH 95b2410e31ebf28b56a4fffffef52c7d13d657ad)
 
 # Get cmake scripts for c4core
 vcpkg_download_distfile(
     CMAKE_ARCHIVE
     URLS "https://github.com/biojppm/cmake/archive/${CM_COMMIT_HASH}.zip"
```
Comment (BillyONeal) [inline ports/c4core/portfile.cmake:19]: Stupid question: Should this use `vcpkg_from_github`?

File: ports/ryml/cmake-fix.patch [hunk 2]
```diff
@@ -25,3 +25,15 @@ index d18407c..db19e0b 100644
...
  if(RYML_WITH_TAB_TOKENS)
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index f6c18ce..7ca8f0b 100644
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -76,7 +76,6 @@ endif()
+
+ c4_install_target(ryml)
+ c4_install_exports(DEPENDENCIES c4core)
+-c4_pack_project()
```
Comment (BillyONeal) [inline ports/ryml/cmake-fix.patch:36]: Can you explain what this patch is doing?

**Review by BillyONeal — CHANGES_REQUESTED** [review]
(Nitpicks already posted, mostly looks OK thanks!)

File: ports/ryml/cmake-fix.patch (hunk 2 above)
Comment (jimwang118) [reply ports/ryml/cmake-fix.patch]: This is to solve the problem of git not being found and has been revised.

File: ports/c4core/portfile.cmake [hunk 3]
```diff
@@ -7,60 +7,38 @@ vcpkg_from_github(
...
-)
-
-vcpkg_extract_source_archive(
-    SOURCE_PATH_FAST_FLOAT 
-    ARCHIVE ${FAST_FLOAT_ARCHIVE}
-    WORKING_DIRECTORY "${CURRENT_BUILDTREES_DIR}/src/deps"
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH_FAST_FLOAT
+    REPO fastfloat/fast_float
+    REF 052975dd5f8166d0f9e4a215fa75a349d5985b91
```
Comment (dg0yt) [inline ports/c4core/portfile.cmake:38]: FTR we do have a fast-float port, but this REF is three years old.

File: ports/ryml/fix_no_find_git.patch [hunk 4]
```diff
@@ -0,0 +1,13 @@
+diff --git a/c4Project.cmake b/c4Project.cmake
+index 60c8717..22ea7fd 100644
+--- a/c4Project.cmake
++++ b/c4Project.cmake
+@@ -1039,7 +1039,7 @@ function(c4_proj_get_version dir tag_o full_o major_o minor_o patch_o tweak_o)
+     if("${dir}" STREQUAL "")
+         set(dir ${CMAKE_CURRENT_LIST_DIR})
+     endif()
+-    find_program(GIT git REQUIRED)
++    find_program(GIT git)
```
Comment (dg0yt) [inline ports/ryml/fix_no_find_git.patch:10]: Hm, I would just let the portfile do something like
~~~cmake
vcpkg_find_acquire_program(GIT)
vcpkg_cmake_configure(... OPTIONS "-DGIT=${GIT}" ...)
~~~
or patch in a way that injects `${VERSION}`.

File: ports/c4core/portfile.cmake (hunk 3 above)
Comment (jimwang118) [reply ports/c4core/portfile.cmake:38]: This is a version requirement for `fast_float` that the upstream depends on. The corresponding `fast_float` version is required in `c4core` version 0.1.11.

**Pull request closed without merging**